File: examples/python/catalog_admin_example.py
#!/usr/bin/env python3

import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO_ROOT / "client/python"))
import uuid

import grpc

from payload_manager_client import PayloadClient
from payload.manager.catalog.v1 import lineage_pb2
from payload.manager.core.v1 import placement_pb2, policy_pb2
from payload.manager.runtime.v1 import lifecycle_pb2, tiering_pb2


def main() -> int:
    # Endpoint can be overridden to point this walkthrough at any running
    # Payload Manager instance.
    target = sys.argv[1] if len(sys.argv) > 1 else "localhost:50051"
    client = PayloadClient(grpc.insecure_channel(target))

    # Allocate a short-lived RAM payload (TTL 60s) so we can safely exercise
    # admin APIs without leaving long-term test artifacts.
    writable = client.AllocateWritableBuffer(16, placement_pb2.TIER_RAM, ttl_ms=60_000, persist=False)
    writable.mmap_obj[:] = bytes(range(1, 17))

    raw_uuid = bytes(writable.descriptor.id.value)
    payload_uuid = uuid.UUID(bytes=raw_uuid)
    client.CommitPayload(str(payload_uuid))
    client.Resolve(str(payload_uuid))

    # Promotion/spill requests demonstrate tiering controls. BEST_EFFORT means
    # the server may partially succeed depending on current constraints.
    promote_request = tiering_pb2.PromoteRequest(target_tier=placement_pb2.TIER_RAM, policy=policy_pb2.PROMOTION_POLICY_BEST_EFFORT)
    promote_request.id.value = raw_uuid
    client.Promote(promote_request)

    spill_request = tiering_pb2.SpillRequest(policy=tiering_pb2.SPILL_POLICY_BEST_EFFORT, wait_for_leases=True)
    spill_request.ids.add().value = raw_uuid
    client.Spill(spill_request)

    # Record a synthetic lineage edge to show how transformation provenance can
    # be attached to payload IDs.
    add_lineage_request = lineage_pb2.AddLineageRequest()
    add_lineage_request.child.value = raw_uuid
    edge = add_lineage_request.parents.add()
    edge.parent.value = raw_uuid
    edge.operation = "identity"
    edge.role = "demo"
    edge.parameters = "{}"
    client.AddLineage(add_lineage_request)

    get_lineage_request = lineage_pb2.GetLineageRequest(upstream=True, max_depth=1)
    get_lineage_request.id.value = raw_uuid
    lineage = client.GetLineage(get_lineage_request)

    # Force-delete keeps the example idempotent and avoids leaving stale
    # payloads behind after the demo completes.
    delete_request = lifecycle_pb2.DeleteRequest(force=True)
    delete_request.id.value = raw_uuid
    client.Delete(delete_request)

    print(f"Catalog/Admin API calls completed for payload {payload_uuid} (lineage edges returned={len(lineage.edges)})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/cpp/catalog_admin_example.cpp
#include <grpcpp/create_channel.h>
#include <grpcpp/security/credentials.h>

#include <cstdint>
#include <iomanip>
#include <iostream>
#include <memory>
#include <sstream>
#include <string>

#include "client/cpp/payload_manager_client.h"
#include "payload/manager/v1.hpp"

namespace {

std::string UuidToHex(const std::string& uuid_bytes) {
  std::ostringstream os;
  for (size_t i = 0; i < uuid_bytes.size(); ++i) {
    if (i == 4 || i == 6 || i == 8 || i == 10) {
      os << '-';
    }
    os << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(static_cast<unsigned char>(uuid_bytes[i]));
  }
  return os.str();
}


} // namespace

int main(int argc, char** argv) {
  // Optional target parameter allows this admin walkthrough to run against any
  // reachable Payload Manager endpoint.
  const std::string target = argc > 1 ? argv[1] : "localhost:50051";

  payload::manager::client::PayloadClient client(grpc::CreateChannel(target, grpc::InsecureChannelCredentials()));

  // Allocate a short-lived RAM payload to exercise tiering/catalog APIs
  // without leaving persistent demo data behind.
  auto writable = client.AllocateWritableBuffer(16, payload::manager::v1::TIER_RAM, 60'000, false);
  if (!writable.ok()) {
    std::cerr << "AllocateWritableBuffer failed: " << writable.status().ToString() << '\n';
    return 1;
  }

  auto writable_payload = writable.ValueOrDie();
  for (int i = 0; i < 16; ++i) {
    writable_payload.buffer->mutable_data()[i] = static_cast<uint8_t>(i + 1);
  }

  const auto& payload_id = writable_payload.descriptor.id();
  const auto  uuid_text  = UuidToHex(payload_id.value());

  auto commit_status = client.CommitPayload(payload_id);
  if (!commit_status.ok()) {
    std::cerr << "CommitPayload failed: " << commit_status.ToString() << '\n';
    return 1;
  }

  auto snapshot = client.Resolve(payload_id);
  if (!snapshot.ok()) {
    std::cerr << "Resolve failed: " << snapshot.status().ToString() << '\n';
    return 1;
  }

  // Promotion and spill operations demonstrate runtime tier-management APIs.
  payload::manager::v1::PromoteRequest promote_request;
  *promote_request.mutable_id() = payload_id;
  promote_request.set_target_tier(payload::manager::v1::TIER_RAM);
  promote_request.set_policy(payload::manager::v1::PROMOTION_POLICY_BEST_EFFORT);

  auto promote = client.Promote(promote_request);
  if (!promote.ok()) {
    std::cerr << "Promote failed: " << promote.status().ToString() << '\n';
    return 1;
  }

  payload::manager::v1::SpillRequest spill_request;
  *spill_request.add_ids() = payload_id;
  spill_request.set_policy(payload::manager::v1::SPILL_POLICY_BEST_EFFORT);
  spill_request.set_wait_for_leases(true);

  auto spill = client.Spill(spill_request);
  if (!spill.ok()) {
    std::cerr << "Spill failed: " << spill.status().ToString() << '\n';
    return 1;
  }

  // Attach a synthetic lineage edge to show provenance graph updates.
  payload::manager::v1::AddLineageRequest add_lineage_request;
  *add_lineage_request.mutable_child() = payload_id;
  auto* parent_edge                    = add_lineage_request.add_parents();
  *parent_edge->mutable_parent()       = payload_id;
  parent_edge->set_operation("identity");
  parent_edge->set_role("demo");
  parent_edge->set_parameters("{}", 2);

  auto add_lineage_status = client.AddLineage(add_lineage_request);
  if (!add_lineage_status.ok()) {
    std::cerr << "AddLineage failed: " << add_lineage_status.ToString() << '\n';
    return 1;
  }

  payload::manager::v1::GetLineageRequest get_lineage_request;
  *get_lineage_request.mutable_id() = payload_id;
  get_lineage_request.set_upstream(true);
  get_lineage_request.set_max_depth(1);

  auto lineage = client.GetLineage(get_lineage_request);
  if (!lineage.ok()) {
    std::cerr << "GetLineage failed: " << lineage.status().ToString() << '\n';
    return 1;
  }

  // Force delete ensures example reruns are clean and deterministic.
  payload::manager::v1::DeleteRequest delete_request;
  *delete_request.mutable_id() = payload_id;
  delete_request.set_force(true);

  auto delete_status = client.Delete(delete_request);
  if (!delete_status.ok()) {
    std::cerr << "Delete failed: " << delete_status.ToString() << '\n';
    return 1;
  }

  std::cout << "Catalog/Admin API calls completed for payload " << uuid_text << " (lineage edges returned=" << lineage.ValueOrDie().edges_size() << ")\n";
  return 0;
}
